In [1]:
#Importing necessary libraries
import pandas as pd
import numpy as np


Reading the Dataframe

In [2]:
df = pd.read_csv("casestudy.csv")

Exploring the dataset

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 685927 entries, 0 to 685926
Data columns (total 4 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Unnamed: 0      685927 non-null  int64  
 1   customer_email  685927 non-null  object 
 2   net_revenue     685927 non-null  float64
 3   year            685927 non-null  int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 20.9+ MB


In [4]:
df.head()

Unnamed: 0         customer_email  net_revenue  year
0           0   nhknapwsbx@gmail.com       249.92  2015
1           1   joiuzbvcpn@gmail.com        87.61  2015
2           2   ukkjctepxt@gmail.com       168.38  2015
3           3   gykatilzrt@gmail.com        62.40  2015
4           4   mmsgsrtxah@gmail.com        43.08  2015

Revenue of the current year


In [5]:
df['year'].value_counts()

2017    249987
2015    231294
2016    204646
Name: year, dtype: int64

In [6]:
#Current year is 2017 . So calculating the net revenue of the current year
df.groupby(['year'])[['net_revenue']].sum()

net_revenue
year             
2015  29036749.19
2016  25730943.59
2017  31417495.03

Revenue of New customers

In [7]:
#finding the count of unique customer emails
customer_array = np.array(np.unique(df['customer_email'], return_counts=True)).T

In [8]:
#Converting the above values to a unique dataframe with 2 columns.
df1 = pd.DataFrame(customer_array, columns = ['Column_A','Column_B'])

In [9]:
df1.head()

Column_A Column_B
0   aaafxtkgxo@gmail.com        1
1   aaagldjgdt@gmail.com        1
2   aaaglfuvti@gmail.com        1
3   aaaicvtnzx@gmail.com        1
4   aaaiekepot@gmail.com        1

In [12]:
#dropping emails with count of 1
df1 = df1.loc[df1["Column_B"] == 1]

In [14]:
#creating a new dataframe from existing dataframe with unique mails
df_customer = df[df['customer_email'].isin(df1['Column_A'])]
df_customer.describe()

Unnamed: 0    net_revenue           year
count  529471.000000  529471.000000  529471.000000
mean   363682.823040     125.691451    2016.108255
std    202893.586372      71.839520       0.861541
min         0.000000       1.000000    2015.000000
25%    187621.500000      63.570000    2015.000000
50%    378693.000000     125.810000    2016.000000
75%    543300.500000     187.910000    2017.000000
max    685926.000000     250.000000    2017.000000

In [15]:
df_customer.groupby(['year'])[['net_revenue']].sum()


net_revenue
year              
2015  2.147600e+07
2016  1.639737e+07
2017  2.867661e+07

New Customer Revenue is 2.8677661e+07

***Existing Customer Growth.***

In [16]:
#defining df with customer_email and their frequency counts
df2 = pd.DataFrame(customer_array, columns = ['Column_A','Column_B'])

In [17]:
df2.describe()

Column_A  Column_B
count                 604618    604618
unique                604618         3
top     bznasflwcl@gmail.com         1
freq                       1    529471

In [21]:
#dropping email with count equal to 1
#it makes sure that only existing customers are included in the calculation of reveneue
df2 = df2.loc[df2["Column_B"] != 1]

In [22]:
df2['Column_B'].value_counts()

2    68985
3     6162
Name: Column_B, dtype: int64

In [23]:
#creating another dataframe with datas from year 2017
dfcurrent = df[df['year']==2017]

In [26]:
#dropping the new customers from year 2017
dfcurrent = dfcurrent[dfcurrent['customer_email'].isin(df1['Column_A'])]

In [28]:
dfcurrent.describe()

Unnamed: 0    net_revenue      year
count  228262.000000  228262.000000  228262.0
mean   563109.774408     125.630230    2017.0
std     72901.987265      71.880915       0.0
min    435940.000000       1.000000    2017.0
25%    499848.250000      63.430000    2017.0
50%    563634.500000     125.740000    2017.0
75%    627452.750000     187.880000    2017.0
max    685926.000000     250.000000    2017.0

In [29]:
#finally collecting the exisiting customers from previous year
current = df[df['customer_email'].isin(dfcurrent['customer_email'])]

In [30]:
current.describe()

Unnamed: 0    net_revenue      year
count  228262.000000  228262.000000  228262.0
mean   563109.774408     125.630230    2017.0
std     72901.987265      71.880915       0.0
min    435940.000000       1.000000    2017.0
25%    499848.250000      63.430000    2017.0
50%    563634.500000     125.740000    2017.0
75%    627452.750000     187.880000    2017.0
max    685926.000000     250.000000    2017.0

In [31]:
current.groupby(['year'])[['net_revenue']].sum()

net_revenue
year             
2017  28676607.64

***Revenue lost from attrition***

In [32]:
#to find the revenue lost from attrition we have to see how many customers we lost in 2017
#converting the unique mails and counts to a data frame
df3 = pd.DataFrame(customer_array, columns = ['Column_A','Column_B'])

In [34]:
#dropping email with count more than 1
df3 = df3.loc[df3["Column_B"] == 1]

In [35]:
#creating another dataframe with datas from year 2016
dfprev = df[df['year']==2016]

In [36]:
#finally collecting the unique customers in year 2016
prev = dfprev[dfprev['customer_email'].isin(df1['Column_A'])]

In [37]:
prev['net_revenue'].sum()

16397370.16

In [38]:
#creating another dataframe with datas from year 2015
dfold = df[df['year']==2015]

In [39]:
#finally collecting the unique customers in year 2015
old = dfold[dfold['customer_email'].isin(df1['Column_A'])]

In [40]:
old['net_revenue'].sum()

21476000.699999996

So the total revenue lost from attrition will be the sum of revenue from  2015 and revenue from 2016, which is equal to 16397370.16+21476000 = 37,873,370

***Existing Customer Revenue Current Year***

While calculating the exisitng customer revenue growth, I have calculated the exisitng customer revenue current year which is 2740887.39

***Existing Customer Revenue Prior Year***

While calculating the exisitng customer revenue growth, I have calculated the exisitng customer revenue in year 2016 which is 2620648.65 and the customer revenue in year 2015 which is 868493.65

***Total Customers Current Year***

In [41]:
#dataframe with datas from year 2017
dfcurrent = df[df['year']==2017]

In [42]:
dfcurrent.describe()

Unnamed: 0    net_revenue      year
count  249987.000000  249987.000000  249987.0
mean   560933.000000     125.676515    2017.0
std     72165.175209      71.924943       0.0
min    435940.000000       1.000000    2017.0
25%    498436.500000      63.390000    2017.0
50%    560933.000000     125.810000    2017.0
75%    623429.500000     187.970000    2017.0
max    685926.000000     250.000000    2017.0

From the above data we can see that the count is 249987 which is the number of customers for current year

***Total Customers Previous Year***

In [44]:
#dataframe with datas from year 2016
dfprev = df[df['year']==2016]

In [45]:
dfprev.describe()

Unnamed: 0    net_revenue      year
count  204646.000000  204646.000000  204646.0
mean   333616.500000     125.733919    2016.0
std     59076.355932      71.838149       0.0
min    231294.000000       1.000000    2016.0
25%    282455.250000      63.722500    2016.0
50%    333616.500000     125.760000    2016.0
75%    384777.750000     187.890000    2016.0
max    435939.000000     250.000000    2016.0

From the above data we can see that the count is 204646 which is the number of customers for year 2016

In [47]:
#dataframe with datas from year 2015
dfold = df[df['year']==2015]

In [48]:
dfold.describe()

Unnamed: 0    net_revenue      year
count  231294.000000  231294.000000  231294.0
mean   115646.500000     125.540434    2015.0
std     66768.970918      71.793027       0.0
min         0.000000       1.000000    2015.0
25%     57823.250000      63.430000    2015.0
50%    115646.500000     125.655000    2015.0
75%    173469.750000     187.560000    2015.0
max    231293.000000     250.000000    2015.0

From the above data we can see that the count is 231294 which is the number of customers for year 2015

***New Customers***

In second part of this case study we calculated the new customer revenue using newdf dataframe. So I am using that df to find the new customers

In [50]:
#dataframe with newcustomers from year 2017
newcustomers = df[df['year']==2017]

In [51]:
newcustomers['customer_email']

435940     mwrossuukz@gmail.com
435941    qvjfterwnk@gmail.com 
435942     vlyigtgfzs@gmail.com
435943     yfirychuhk@gmail.com
435944     fdkiioqtli@gmail.com
                  ...          
685922     qzqttwiftu@gmail.com
685923     pjodiifjop@gmail.com
685924     appaplmgko@gmail.com
685925     wvkpmwsgck@gmail.com
685926     aregboumbw@gmail.com
Name: customer_email, Length: 249987, dtype: object

The customer mails belong to the new customers, and the total new customers is 228262

***Lost Customers***

In [53]:
customers2016 = df[df['year']==2016]

In [54]:
customers2016.describe()

Unnamed: 0    net_revenue      year
count  204646.000000  204646.000000  204646.0
mean   333616.500000     125.733919    2016.0
std     59076.355932      71.838149       0.0
min    231294.000000       1.000000    2016.0
25%    282455.250000      63.722500    2016.0
50%    333616.500000     125.760000    2016.0
75%    384777.750000     187.890000    2016.0
max    435939.000000     250.000000    2016.0

In [55]:
customers2017 = df[df['year']==2017]

In [56]:
#calculating the exisiting customers in 2017 from 2016
existcustomers = customers2016[customers2016['customer_email'].isin(customers2017['customer_email'])]

In [57]:
existcustomers.describe()

Unnamed: 0   net_revenue     year
count   20959.000000  20959.000000  20959.0
mean   333325.722840    125.036913   2016.0
std     59288.447226     71.551989      0.0
min    231301.000000      1.010000   2016.0
25%    281861.500000     63.315000   2016.0
50%    332788.000000    124.630000   2016.0
75%    384970.000000    185.910000   2016.0
max    435937.000000    249.970000   2016.0

We can see that out of 204646 customers from 2016, only 20959 exist in 2017. Therefore the lost customers = 183687